db/db_impl/db_impl_open.cc

```diff
@@ -1929,4 +1929,305 @@ Status DBImpl::Open(const DBOptions& db_options, const std::string& dbname,
+                f->file_checksum_func_name, f->min_timestamp, f->max_timestamp);
+          }
+        }
+      }
+      if (!s.ok()) {
+        break;
+      }
+    }
+    new_db_impl->versions_->FetchAddFileNumber(next_file_number);
+  }
+  // Merge memtable.
+  if (s.ok()) {
+    if (mopts.merge_memtable) {
+      uint64_t max_seq_number = 0;
+      autovector<MemTable*> to_delete;  // not used.
+      InstrumentedMutexLock lock(&new_db_impl->mutex_);
+      for (auto* db : db_impls) {
+        // Block all writes.
+        db->mutex_.Lock();
+        max_seq_number =
+            std::max(max_seq_number, db->versions_->LastSequence());
+      }
+      new_db_impl->versions_->SetLastAllocatedSequence(max_seq_number);
+      new_db_impl->versions_->SetLastSequence(max_seq_number);
+      for (size_t i = 0; i < cf_ids.size(); i++) {
+        auto& cf_name = column_families[i].name;
+        auto new_cfd = new_cfds[i];
+        auto& mutable_cf_options = *new_cfd->GetLatestMutableCFOptions();
+        autovector<MemTable*> mems;
+        for (auto* db : db_impls) {
+          auto cfd =
+              db->versions_->GetColumnFamilySet()->GetColumnFamily(cf_name);
+          assert(cfd != nullptr && !cfd->IsDropped());
+          cfd->imm()->ExportMemtables(&mems);
+          if (!cfd->mem()->IsEmpty()) {
+            mems.push_back(cfd->mem());
+          }
+        }
+        auto* fresh_mem =
+            new_cfd->ConstructNewMemtable(mutable_cf_options, max_seq_number);
+        mems.push_back(fresh_mem);
+        for (auto mem : mems) {
+          assert(mem != nullptr);
+          new_cfd->imm()->Add(new_cfd->mem(), &to_delete);
+          mem->Ref();
+          new_cfd->SetMemtable(mem);
+        }
+      }
+      for (auto* db : db_impls) {
+        db->mutex_.Unlock();
+      }
+    } else {
+      uint64_t max_seq_number = 0;
+      for (auto* db : db_impls) {
+        if (db->write_buffer_manager_->buffer_size() > 0) {
+          s = Status::InvalidArgument("DB has unpersisted data.");
+          break;
+        }
+        max_seq_number =
+            std::max(max_seq_number, db->versions_->LastSequence());
+      }
+      new_db_impl->versions_->SetLastAllocatedSequence(max_seq_number);
+      new_db_impl->versions_->SetLastSequence(max_seq_number);
+    }
+  }
+  // Apply version edits.
+  if (s.ok()) {
+    InstrumentedMutexLock lock(&new_db_impl->mutex_);
+    for (size_t i = 0; s.ok() && i < cf_ids.size(); i++) {
+      auto cfd = new_cfds[i];
+      auto& mutable_cf_options = *cfd->GetLatestMutableCFOptions();
+      SuperVersionContext sv_context(/* create_superversion */ true);
+      s = new_db_impl->versions_->LogAndApply(
+          cfd, mutable_cf_options, &cf_edits[i], &new_db_impl->mutex_, nullptr,
+          false);
+      new_db_impl->InstallSuperVersionAndScheduleWork(cfd, &sv_context,
+                                                      mutable_cf_options);
+      sv_context.Clean();
+    }
+  }
+  // Cleaning up.
```

Comment:
should the new merged db's files be deleted here if !s.ok() ?

Reply:
I guess user has to manually `DestroyDB`. This behavior is consistent with `DB::Open`.